### PR TITLE
New API: Port documentLink

### PR DESCRIPTION
**Pull request by rubenporras**

### CODE REVIEW

**Comment by rubenporras** [comment]
@ahmedneilhussain , looks good to you?

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/documentLink/DocumentLinkDetector.java [hunk 1]
```diff
@@ -80,35 +80,18 @@ public IHyperlink[] detectHyperlinks(ITextViewer textViewer, IRegion region, boo
...
-										return null;
-									} catch (TimeoutException e) {
-										LanguageServerPlugin.logWarning("Could not detect hyperlinks due to timeout after 2 seconds in `document/Link`", e); //$NON-NLS-1$
-										return null;
-									}
-								}).filter(Objects::nonNull).flatMap(List<DocumentLink>::stream).map(link -> {
+			return LanguageServers.forDocument(document)
+					.withFilter(capabilities -> capabilities.getDocumentLinkProvider() != null)
+					.collectAll(languageServer -> languageServer.getTextDocumentService().documentLink(params))
+					.thenApplyAsync(links -> {
```
Comment (ahmedneilhussain) [inline org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/documentLink/DocumentLinkDetector.java:86]: `async` not needed - already done by `collectAll`.

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/documentLink/DocumentLinkDetector.java [hunk 2]
```diff
@@ -80,35 +80,18 @@ public IHyperlink[] detectHyperlinks(ITextViewer textViewer, IRegion region, boo
...
+						IHyperlink[] res = links.stream().flatMap(List<DocumentLink>::stream).filter(Objects::nonNull)
+								.filter(link -> link.getTarget() != null).map(link -> {
```
Comment (ahmedneilhussain) [inline org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/documentLink/DocumentLinkDetector.java:88]: Not your doing, but this is a bit of a monster lambda expression.

Could you pull some of it out into a method while you're in the code? Maybe a (lsp4j) DocumentLink -> (eclipse) DocumentHyperLink converter?

File: org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/documentLink/LSPDocumentLinkPresentationReconcilingStrategy.java [hunk 3]
```diff
@@ -70,18 +70,17 @@ private void underline() {
...
-							if (control != null) {
-								control.getDisplay().asyncExec(() -> underline(links));
-							}
-						})).toArray(CompletableFuture[]::new)));
+		cancel();
+		final var params = new DocumentLinkParams(LSPEclipseUtils.toTextDocumentIdentifier(uri));
+		request = LanguageServers.forDocument(document)
+				.withFilter(capabilities -> capabilities.getDocumentLinkProvider() != null)
+				.collectAll(languageServer -> languageServer.getTextDocumentService().documentLink(params))
+				.thenAcceptAsync(links -> {
```
Comment (ahmedneilhussain) [inline org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/documentLink/LSPDocumentLinkPresentationReconcilingStrategy.java:78]: Again, async not needed .... unless using the display as an executor

Comment (ahmedneilhussain) [reply org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/documentLink/LSPDocumentLinkPresentationReconcilingStrategy.java:78]: .. which you can't do if control is null. Although whether that is something that should be guarded against is debatable. Could extract the display at the top of this method and then use it as the async executor.

**Comment by ahmedneilhussain** [comment]
Couple of points, but LGTM broadly.

**Comment by rubenporras** [comment]
@ahmedneilhussain , good now? 

Yes, I knew about the big method, but I wanted to keep my PR as small as possible. Now it is part of a second commit inside the PR.

**Comment by ahmedneilhussain** [comment]
> @ahmedneilhussain , good now?
> 
> Yes, I knew about the big method, but I wanted to keep my PR as small as possible. Now it is part of a second commit inside the PR.

Yeah, thanks. I think you're right to separate out the commits and keep the API changes distinct from unrelated (but welcome) code improvements.

**Comment by rubenporras** [comment]
I will merge this as well, if comments come, I will address them as well

**Comment by sebthom** [comment]
@rubenporras did you on purpose increase the timeout from 2 to 4 seconds in the detectHyperlinks method? the problem is that this runs in the UI thread and freezes the GUI. and with slow language servers it freezes a lot.

**Comment by rubenporras** [comment]
I do not know for sure, but most probably intentional. We have some cases where our LS is not so fast, and we prefer it to be slower than not to present any information.
I think we should make this timeouts configurable.

**Pull request merged**